### PR TITLE
Do not rely on JavaConverters in our implementation of java.util.

### PR DESCRIPTION
**Pull request by sjrd**
Instead of using some compatibility hacks to access the Java collection converters across all versions of Scala, we simply get rid of them. Instead, we implement our own small set of combinators for operations that we actually use. We use the opportunity to make those combinators zero-cost.

This PR is prompted by yet another breaking change to the API of `JavaConverters` coming in 2.13.x. See https://github.com/scala/scala/pull/7987#discussion_r278158104

### CODE REVIEW

File: javalib/src/main/scala/java/util/AbstractList.scala [hunk 1]
```diff
@@ -52,8 +52,8 @@ abstract class AbstractList[E] protected () extends AbstractCollection[E]
 
   def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
     checkIndexOnBounds(index)
-    for ((elem, i) <- c.iterator.asScala.zipWithIndex)
-      add(index + i, elem)
+    for (elemAndIdx <- c.iterator().scalaOps.zipWithIndex.scalaOps)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/AbstractList.scala]: Why not just add `withFilter`?

File: javalib/src/main/scala/java/util/Collections.scala [hunk 2]
```diff
@@ -249,14 +249,14 @@ object Collections {
     min(coll, naturalComparator[T])
 
   def min[T](coll: Collection[_ <: T], comp: Comparator[_ >: T]): T =
-    coll.asScala.min(comp)
+    coll.scalaOps.reduceLeft((a, b) => if (comp.compare(a, b) <= 0) a else b)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Collections.scala]: `reduceLeft(math.min)`?

File: javalib/src/main/scala/java/util/Collections.scala [hunk 3]
```diff
@@ -249,14 +249,14 @@ object Collections {
...
 
   // Differs from original type definition, original: [T <: jl.Comparable[_ >: T]]
   def max[T <: AnyRef with jl.Comparable[T]](coll: Collection[_ <: T]): T =
     max(coll, naturalComparator[T])
 
   def max[T](coll: Collection[_ <: T], comp: Comparator[_ >: T]): T =
-    coll.asScala.max(comp)
+    coll.scalaOps.reduceLeft((a, b) => if (comp.compare(a, b) >= 0) a else b)
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Collections.scala]: `reduceLeft(math.max)`?

File: javalib/src/main/scala/java/util/Collections.scala [hunk 4]
```diff
@@ -534,22 +534,41 @@ object Collections {
...
+      !c1.scalaOps.exists(elem => c2.contains(elem))
     else
-      !c2.asScala.exists(elem => c1.contains(elem))
+      !c2.scalaOps.exists(elem => c1.contains(elem))
   }
 
-  def addAll[T](c: Collection[_ >: T], elements: Array[AnyRef]): Boolean =
-    c.addAll(elements.asInstanceOf[Array[T]].toSeq.asJava)
+  def addAll[T](c: Collection[_ >: T], elements: Array[AnyRef]): Boolean = {
+    val elementsColl = new AbstractCollection[T] {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Collections.scala]: Huh, couldn't you use standard array ops to make an Scala iterator out of this and then wrap it as a java iterator?

File: javalib/src/main/scala/java/util/Hashtable.scala [hunk 5]
```diff
@@ -87,25 +87,87 @@ class Hashtable[K, V] private (inner: mutable.HashMap[Box[Any], V])
...
+
+  private class EntrySetIterator extends Iterator[Map.Entry[K, V]] {
+    private var underlying: scala.collection.Iterator[(Box[Any], V)] =
+      Hashtable.this.inner.iterator
+
+    private var isSnapshot: Boolean = false
+
+    private var prev: Box[Any] = null
+
+    private def ensureSnapshot(): Unit = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Hashtable.scala]: I assume this logic is copied from Scala itself?

File: javalib/src/main/scala/java/util/Properties.scala [hunk 6]
```diff
@@ -47,19 +49,16 @@ class Properties(protected val defaults: Properties)
...
-    val thisSet = keySet().asScala.map(_.asInstanceOf[String])
-    val defaultsIterator =
-      if (defaults != null) defaults.propertyNames().asScala
-      else scala.collection.Iterator.empty
-    val filteredDefaults = defaultsIterator.collect {
-      case k: String if !thisSet(k) => k
-    }
-    (thisSet.iterator ++ filteredDefaults).asJavaEnumeration
+    val propNames = mutable.Set.empty[Any]
+    keySet().scalaOps.foreach(propNames += _.asInstanceOf[String]) // CCE mandated by spec
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Properties.scala]: Consider writing CCE out for ease of understanding.

File: javalib/src/main/scala/java/util/Collections.scala (hunk 2 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Collections.scala]: That would fail to use the supplied `comp` comparator. (Also `math.min` only works for private numbers, not for any `T`.)

File: javalib/src/main/scala/java/util/Collections.scala (hunk 3 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Collections.scala]: That would fail to use the supplied `comp` comparator. (Also `math.max` only works for private numbers, not for any `T`.)

File: javalib/src/main/scala/java/util/Hashtable.scala (hunk 5 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Hashtable.scala]: More or less, but there are adaptations. I have added a comment.

File: javalib/src/main/scala/java/util/Hashtable.scala [hunk 7]
```diff
@@ -87,25 +87,96 @@ class Hashtable[K, V] private (inner: mutable.HashMap[Box[Any], V])
...
     }
-    inner.asJava.entrySet().asScala.map(new UnboxedEntry(_): ju.Map.Entry[K, V]).asJava
+
+    def remove(): Unit = {
+      if (prev == null)
+        throw new IllegalStateException("next must be called at least once before remove")
+      ensureSnapshot()
+      inner -= prev
+      prev = null
+    }
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Hashtable.scala]: It seems this codepath is completely untested or am I missing something?

Comment (sjrd) [reply javalib/src/main/scala/java/util/Hashtable.scala]: I added a test for the corner cases of the entrySet's iterator.
